### PR TITLE
Fixed #32458 -- Made __repr__() for Index and BaseConstraint subclasses more consistent.

### PR DESCRIPTION
**Pull request by hannseman**
Ticket: https://code.djangoproject.com/ticket/32458

### CODE REVIEW

File: django/contrib/postgres/constraints.py [hunk 1]
```diff
@@ -155,12 +155,15 @@ def __eq__(self, other):
...
-            '' if self.condition is None else ', condition=%s' % self.condition,
-            '' if self.deferrable is None else ', deferrable=%s' % self.deferrable,
-            '' if not self.include else ', include=%s' % repr(self.include),
-            '' if not self.opclasses else ', opclasses=%s' % repr(self.opclasses),
+            self.name,
+            "'%s'" % ', '.join([
+                str(expression) for expression in self.expressions
+            ]),
+            '' if self.condition is None else " condition='%s'" % self.condition,
+            '' if self.deferrable is None else " deferrable='%s'" % self.deferrable,
```
Comment (felixxm) [inline django/contrib/postgres/constraints.py:166]: I would revert these changes, a string representation of `condition` and `deferrable` doesn't need and extra quotes.

File: django/contrib/postgres/constraints.py [hunk 2]
```diff
@@ -155,12 +155,15 @@ def __eq__(self, other):
...
             self.index_type,
-            self.expressions,
-            '' if self.condition is None else ', condition=%s' % self.condition,
-            '' if self.deferrable is None else ', deferrable=%s' % self.deferrable,
-            '' if not self.include else ', include=%s' % repr(self.include),
-            '' if not self.opclasses else ', opclasses=%s' % repr(self.opclasses),
+            self.name,
+            "'%s'" % ', '.join([
+                str(expression) for expression in self.expressions
+            ]),
```
Comment (felixxm) [inline django/contrib/postgres/constraints.py:164]: After checking changes in tests I think using `repr()` for `expressions`, `fields`, `include`, and `opclasses` makes a representation more readable (it's easy to miss a comma and don't notice the list of values), e.g.
```python
' expressions=%s' % repr(self.expressions)
' fields=%s' % repr(self.fields)
' opclasses=%s' % repr(self.opclasses)
' include=%s' % repr(self.include)
```

**Review by felixxm — COMMENTED** [review]
@hannseman Thanks :+1:

File: django/contrib/postgres/constraints.py (hunk 1 above)
Comment (hannseman) [reply django/contrib/postgres/constraints.py:166]: Felt a bit weird to leave those unquoted as all other arguments was quoted but it makes a better fit with your proposal of running repr on sequences (much better). 👍

**Comment by hannseman** [comment]
I also made `CheckConstraint` follow the same format where `check` isn't quoted. I made sure to run `str` and `repr` before %-formatting to make sure that we don't crash on bad user input, for example someone passes a tuple as `name` and `repr` blows up on `self.name = "abc",` and  `'%s % self.name`. We could also wrap the values in tuples but IMO this is more readable.

File: django/contrib/postgres/constraints.py [hunk 3]
```diff
@@ -155,12 +155,13 @@ def __eq__(self, other):
...
-            self.expressions,
-            '' if self.condition is None else ', condition=%s' % self.condition,
-            '' if self.deferrable is None else ', deferrable=%s' % self.deferrable,
-            '' if not self.include else ', include=%s' % repr(self.include),
-            '' if not self.opclasses else ', opclasses=%s' % repr(self.opclasses),
+            repr(self.index_type),
+            repr(self.name),
+            repr(self.expressions),
+            '' if self.condition is None else ' condition=%s' % str(self.condition),
+            '' if self.deferrable is None else ' deferrable=%s' % str(self.deferrable),
```
Comment (felixxm) [inline django/contrib/postgres/constraints.py]: Type of `condition` and `deferrable` is checked in `__init__()` so `str()` is not necessary.

File: django/db/models/constraints.py [hunk 4]
```diff
@@ -161,10 +161,12 @@ def remove_sql(self, model, schema_editor):
         )
 
     def __repr__(self):
-        return '<%s: fields=%r name=%r%s%s%s%s>' % (
-            self.__class__.__name__, self.fields, self.name,
-            '' if self.condition is None else ' condition=%s' % self.condition,
-            '' if self.deferrable is None else ' deferrable=%s' % self.deferrable,
+        return '<%s: fields=%s name=%s%s%s%s%s>' % (
+            self.__class__.__name__,
```
Comment (felixxm) [inline django/db/models/constraints.py]: We should use `__qualname__` in all classes.

File: django/contrib/postgres/constraints.py [hunk 5]
```diff
@@ -155,12 +155,13 @@ def __eq__(self, other):
...
             self.__class__.__qualname__,
-            self.index_type,
-            self.expressions,
-            '' if self.condition is None else ', condition=%s' % self.condition,
-            '' if self.deferrable is None else ', deferrable=%s' % self.deferrable,
-            '' if not self.include else ', include=%s' % repr(self.include),
-            '' if not self.opclasses else ', opclasses=%s' % repr(self.opclasses),
+            repr(self.index_type),
+            repr(self.name),
+            repr(self.expressions),
```
Comment (felixxm) [inline django/contrib/postgres/constraints.py]: `expressions` should be before the `name` like in other classes.

**Review by felixxm — COMMENTED** [review]
@hannseman Thanks for updates :+1: I pushed small edits.

**Comment by felixxm** [comment]
buildbot, test on oracle.

File: django/db/models/indexes.py [hunk 6]
```diff
@@ -169,15 +169,15 @@ def set_name_with_model(self, model):
...
-            '' if not self.fields else " fields='%s'" % ', '.join(self.fields),
-            '' if not self.expressions else " expressions='%s'" % ', '.join([
-                str(expression) for expression in self.expressions
-            ]),
+        return '<%s:%s%s%s%s%s%s%s>' % (
+            self.__class__.__qualname__,
+            '' if not self.fields else ' fields=%s' % repr(self.fields),
+            '' if not self.expressions else ' expressions=%s' % repr(self.expressions),
+            '' if not self.name else ' name=%s' % repr(self.name),
+            '' if self.db_tablespace is None else ' db_tablespace=%s' % repr(self.db_tablespace),
```
Comment (hannseman) [inline django/db/models/indexes.py:177]: I noticed that `db_tablespace` was missing.

**Comment by felixxm** [comment]
@hannseman Thanks :+1:

**Comment by hannseman** [comment]
@felixxm thanks for the review and fixes!
